### PR TITLE
Loading button

### LINKED ISSUE #4941: Feature suggestion : Loading button
# TL;DR
Have `Component:loading-btn` which display a loading animation in it, +transition with existing `btn` stuff.
_Think it's a good idea ? add :+1:
Thanks ;)_

---

## Expected Behavior
I would like to see this kind of button :
image
See https://github.com/Dogfalo/materialize/pull/4942 for more

## UseCase
- In a form
- You click a *normal* button
- By adding class `is-loading` to it, it displays a loading animation
- Plus disabling button action, avoiding multi-submit & stuff
- Do your async/ajax/WTFYouWant
- Delete `is-loading` class and *YaY it's done*

## Material compliance
- Well, nothing is clearly stated about button & their motion
- But it's stay compliant if we respect [these motion rules](https://material.io/guidelines/motion/material-motion.html)

## Current Behavior
- Well there is no loading `btn` for the moment
- We could edit the content of a button with [mini preloader](http://materializecss.com/preloader.html) but it's not really a good solution as you need to edit DOM and too much stuff on html/css part of things.

## Possible Solution
Testable in [this fork](https://github.com/storm1er/materialize/tree/loading-btn)

### PR DESCRIPTION
**Pull request by storm1er**
## Proposed changes
Will close #4941 

## Screenshots :
image

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue).
- [X] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to change).

## Checklist:
- [X] I have read the **[CONTRIBUTING document](https://github.com/Dogfalo/materialize/blob/master/CONTRIBUTING.md)**.
- [X] My change requires a change to the documentation.
- [X] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes. *needed for overlay !*
- [X] All new and existing tests passed.

### CODE REVIEW

**Comment by storm1er** [comment]
@tomscholz

I'll need help about [icons transition](https://github.com/Dogfalo/materialize/pull/4942/files#diff-886c86476c5a68d98b28f20e0df98b0fR322) ... I wish they could fade away then let loading appear but i don't know how to atm.

Also, I need a moment to create [animation color](https://github.com/Dogfalo/materialize/pull/4942/commits/5d3da193d0f9674d144e27c2a0cfbf36c80c3b3d#diff-886c86476c5a68d98b28f20e0df98b0fR294) ... animation will stay white for the moment, I'll make them compatible with `*-text` colors.

**Comment by Dogfalo** [comment]
Please see my comments on the #4941

**Comment by storm1er** [comment]
Just added overlay + updated description
-> I think overlay should have a Jasmine test

File: jade/page-contents/buttons_content.html [hunk 1]
```diff
@@ -167,16 +167,99 @@ <h2 class="header">Large</h2>
...
+              $(this).toggleClass('is-loading');
+            });
+          });
+        </script>
+        <pre><code class="language-markup col s12">
+&lt;a class="waves-effect waves-light btn-loading with-overlay is-clickable">Clickable&lt;/a>
+&lt;a class="waves-effect waves-light btn-loading-icons with-overlay is-clickable">&lt;i class="material-icons left">cloud&lt;/i>Clickable&lt;/a>
+&lt;script>
+  window.addEventListener("load", function(event) {
+    $('.btn-loading.with-overlay, .btn-loading-icons.with-overlay').on('click', function(){
```
Comment (storm1er) [inline jade/page-contents/buttons_content.html]: missing explicative comment as below

File: js/buttons.js [hunk 2]
```diff
@@ -32,6 +32,11 @@
       FABtoToolbar($menu);
     });
 
+    // enable btn-loading overlays
+    $('.btn-loading-icons.with-overlay,.btn-loading.with-overlay').each(function(index, el) {
```
Comment (storm1er) [inline js/buttons.js]: no way to avoid `.each()` ?

File: js/buttons.js [hunk 3]
```diff
@@ -263,5 +271,93 @@
     }, 200);
   };
 
+  var initOverlayBtn = function(btn) {
+    // buildiing overlay
```
Comment (storm1er) [inline js/buttons.js]: buildiiiiiiiing ?

File: js/buttons.js [hunk 4]
```diff
@@ -263,5 +271,93 @@
...
+    var overlay = $('<div class="button-overlay"></div>');
+    overlay.css({
+      'display': 'none',
+      'z-index': 1000,
+      'width': '0px',
+      'height': '0px',
+      'position': 'fixed',
+      'background-color': 'rgba(0, 0, 0, 0.5)',
+      'transition': 'all ease 700ms'
+    });
```
Comment (storm1er) [inline js/buttons.js]: move css to _buttons.scss ?

File: js/buttons.js [hunk 5]
```diff
@@ -263,5 +271,93 @@
...
+    $('body').append(overlay);
+
+    // keeping user z-index if set
```
Comment (storm1er) [inline js/buttons.js]: z-index of what ?

File: js/buttons.js [hunk 6]
```diff
@@ -263,5 +271,93 @@
...
+    var defaultZIndex = '';
+
+    // listening for .is-loading class
+    var observer = new MutationObserver(function(mutations) {
+      mutations.forEach(function(mutation) {
+        if (mutation.attributeName === "class") {
+          // is-loading
+          if (btn.hasClass('is-loading')) {
+            console.info('loading activated');
```
Comment (storm1er) [inline js/buttons.js]: clean console logs !!!

File: js/buttons.js [hunk 7]
```diff
@@ -263,5 +271,93 @@
...
+                'border-radius': '0px',
+                'width': '100%',
+                'height': '100%',
+                'top': '0px',
+                'left': '0px'
+              });
+            }, 100);
+          // !is-loading
+          } else if (!btn.hasClass('is-loading')) {
+            console.info('loading de-activated');
```
Comment (storm1er) [inline js/buttons.js]: clean console logs !!!

File: js/buttons.js [hunk 8]
```diff
@@ -263,5 +271,93 @@
...
+            // when overlay initial position is set, show it
+            // timer is here too avoid css reflow
+            setTimeout(function(){
+              overlay.css({
+                'border-radius': '0px',
+                'width': '100%',
+                'height': '100%',
+                'top': '0px',
+                'left': '0px'
+              });
```
Comment (storm1er) [inline js/buttons.js]: move css to _buttons.scss ?

File: js/buttons.js [hunk 9]
```diff
@@ -263,5 +271,93 @@
...
+          } else if (!btn.hasClass('is-loading')) {
+            console.info('loading de-activated');
+            // make overlay return too btn
+            overlay.css({
+              'border-radius': '5px',
+              'width': parseInt(btn.innerWidth())+'px',
+              'height': parseInt(btn.innerHeight())+'px',
+              'top': parseInt(btn.offset().top - $(window).scrollTop()) + 'px',
+              'left': parseInt(btn.offset().left) + 'px'
+            });
```
Comment (storm1er) [inline js/buttons.js]: move css to _buttons.scss ?

File: js/buttons.js [hunk 10]
```diff
@@ -263,5 +271,93 @@
...
+
+    // if user delete btn with $(btn).remove();
+    btn.on('destroyed', function(){
+      overlay.remove();
+      observer.disconnect();
+      console.info('overlay removed');
+    });
+  };
+
+  // call $(el).on('destroyed') when $(el).remove() is used
```
Comment (storm1er) [inline js/buttons.js]: Applicable to whole projet, this should be in another file.

File: sass/components/_buttons.scss [hunk 11]
```diff
@@ -289,3 +289,197 @@ button.btn-floating {
...
+}
+@keyframes fadeIn {
+  0% {
+    opacity: 0;
+  }
+  100% {
+    opacity: 1;
+  }
+}
+
```
Comment (storm1er) [inline sass/components/_buttons.scss]: rename animation to btnLoading_fadeOut/In/...

File: sass/components/_buttons.scss [hunk 12]
```diff
@@ -289,3 +289,197 @@ button.btn-floating {
...
+        // because .text-* classes already use !important and make text appear if not present
+        // &.#{$color_name}-text.text-#{$color_type} {
+        //   color: transparent !important;
+        // }
+        &.#{$color_name}-text.text-#{$color_type}:after {
+          border: 2px solid $color_value !important;
+          border-radius: 290486px;
+          border-right-color: transparent;
+          border-top-color: transparent;
+        }
```
Comment (storm1er) [inline sass/components/_buttons.scss]: doesn't seems to work, check z-index maybe ?

File: sass/components/_buttons.scss [hunk 13]
```diff
@@ -289,3 +289,197 @@ button.btn-floating {
...
+        // because .text-* classes already use !important and make text appear if not present
+        // &.#{$color_name}-text.text-#{$color_type} i.material-icons {
+        //   color: transparent !important;
+        // }
+        &.#{$color_name}-text.text-#{$color_type} i.material-icons:after {
+          border: 2px solid $color_value !important;
+          border-radius: 290486px;
+          border-right-color: transparent;
+          border-top-color: transparent;
+        }
```
Comment (storm1er) [inline sass/components/_buttons.scss]: doesn't seems to work, check z-index maybe ?

File: js/buttons.js [hunk 14]
```diff
@@ -263,5 +271,93 @@
...
+            btn.css('z-index', 1001);
+            // initial overlay position
+            overlay.css({
+              'display': 'block',
+              'border-radius': '5px',
+              'width': parseInt(btn.innerWidth())+'px',
+              'height': parseInt(btn.innerHeight())+'px',
+              'top': (btn.offset().top - $(window).scrollTop()) + 'px',
+              'left': btn.offset().left + 'px'
+            });
```
Comment (storm1er) [inline js/buttons.js]: move css to _buttons.scss ?

File: jade/page-contents/buttons_content.html (hunk 1 above)
Comment (storm1er) [inline jade/page-contents/buttons_content.html]: **done**

File: js/buttons.js (hunk 3 above)
Comment (storm1er) [inline js/buttons.js]: **done**

File: js/buttons.js (hunk 4 above)
Comment (storm1er) [inline js/buttons.js]: **done**

File: js/buttons.js (hunk 5 above)
Comment (storm1er) [inline js/buttons.js]: **done**

File: js/buttons.js (hunk 6 above)
Comment (storm1er) [inline js/buttons.js]: **done**

File: js/buttons.js (hunk 7 above)
Comment (storm1er) [inline js/buttons.js]: **done**

File: js/buttons.js (hunk 14 above)
Comment (storm1er) [inline js/buttons.js]: **no, realtime calculation is needed**

File: js/buttons.js (hunk 8 above)
Comment (storm1er) [inline js/buttons.js]: **done**

File: js/buttons.js (hunk 9 above)
Comment (storm1er) [inline js/buttons.js]: **no, realtime calculation is needed**

File: js/buttons.js (hunk 10 above)
Comment (storm1er) [inline js/buttons.js]: @tomscholz or maybe @Dogfalo 
I need some advice on this

How/Where do you define _"global function"_ like this in Materializecss ? 

Is using `js/init.js` a good idea ?

File: sass/components/_buttons.scss (hunk 11 above)
Comment (storm1er) [inline sass/components/_buttons.scss]: **done**

File: sass/components/_buttons.scss (hunk 12 above)
Comment (storm1er) [inline sass/components/_buttons.scss]: **done** just needed to remove `!important`

File: sass/components/_buttons.scss (hunk 13 above)
Comment (storm1er) [inline sass/components/_buttons.scss]: **done** just needed to remove `!important`

**Comment by Dogfalo** [comment]
After some consideration, we don't think that this will be a good fit in the project. It seems that a lot of code needs to be added for this small feature and furthermore I think the more material pattern would actually be to remove the button and show a spinner in its place, not within the button.

**Comment by storm1er** [comment]
> I think the more material pattern would actually be to remove the button and show a spinner in its place, not within the button.

You're probably right one this one. So can we start a new PR with this ?
You can see on #4941 that people want this feature, some has even melted frameworks together like Ladda to get this feature.

**Comment by jpic** [comment]
Of course it is necessary to give feedback on click to the user how can such a feature be rejected given all the work a contributor has put in it ? I wonder.

**Comment by storm1er** [comment]
@jpic Well, there is so much way to make this feature available through so much JS library that it was "implementing something user would probably customize themselves". 
I understand why this hasn't been done.

Anyway, since there is so much commits to catch, you can see what've done in "Files changes" tab and DIY or reopen a new PR ;)

**Pull request closed without merging**